annotation-processor/src/main/java/org/hibernate/validator/ap/classchecks/AbstractMethodOverrideCheck.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.ap.classchecks;
+
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+import org.hibernate.validator.ap.checks.ConstraintCheckIssue;
+import org.hibernate.validator.ap.util.CollectionHelper;
+
+/**
+ * Abstract base class for {@link ClassCheck} implementations that check overridden methods.
+ *
+ * @author Marko Bekhta
+ */
+public abstract class AbstractMethodOverrideCheck extends AbstractClassCheck {
+
+	private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+	private static final String JAVAX_VALIDATION_CONSTRAINT = "javax.validation.Constraint";
+
+	private final Elements elementUtils;
+
+	private final Types typeUtils;
+
+	public AbstractMethodOverrideCheck(Elements elementUtils, Types typeUtils) {
+		this.elementUtils = elementUtils;
+		this.typeUtils = typeUtils;
+	}
+
+	@Override
+	public Collection<ConstraintCheckIssue> checkMethod(ExecutableElement currentMethod) {
+		if ( !needToPerformAnyChecks( currentMethod ) ) {
+			return Collections.emptySet();
+		}
+
+		TypeElement currentTypeElement = getEnclosingTypeElement( currentMethod );
+
+		// if current type is java.lang.Object then we can continue without doing any other checks
+		if ( isJavaLangObjectOrNull( currentTypeElement ) ) {
+			return Collections.emptySet();
+		}
+		// find if there's a method that was overridden by the current one.
+		Map<Boolean, List<ExecutableElement>> overriddenMethods = findAllOverriddenElements( currentTypeElement, currentMethod );
+		if ( overriddenMethods.get( Boolean.TRUE ).isEmpty() && overriddenMethods.get( Boolean.FALSE ).isEmpty() ) {
+			return Collections.emptySet();
+		}
+
+		// if there's more than one overridden originally declared method we need to make sure that all of them match
+		Set<ConstraintCheckIssue> errors = CollectionHelper.newHashSet();
+		for ( ExecutableElement firstExecutable : overriddenMethods.get( Boolean.TRUE ) ) {
```

Comment:
after thinking about it I'd say even though it might be better to have different checks for each rule from a logical point of view but it will save some time (during execution) if we keep both rules for parameter checks in one check class (as for both of the rules inheritance hierarchy is needed)